brave/utility/tor/tor_impl.cc

```diff
@@ -0,0 +1,169 @@
+// Copyright (c) 2018 The Brave Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "brave/utility/tor/tor_impl.h"
+
+#include <utility>
+
+#include "base/command_line.h"
+#include "base/process/kill.h"
+#include "base/process/launch.h"
+#include "base/single_thread_task_runner.h"
+#include "base/task_scheduler/post_task.h"
+#include "base/threading/thread.h"
+#include "base/threading/thread_task_runner_handle.h"
+
+#if defined(OS_POSIX)
+int pipehack[2];
+
+static void SIGCHLDHandler(int signal) {
+  int error = errno;
+  char ch = 0;
+  write(pipehack[1], &ch, 1);
+  errno = error;
+}
+
+static void SetupPipeHack() {
+  if (pipe(pipehack) == -1)
+    LOG(ERROR) << "pipehack";
+
+  int flags;
+  for (size_t i = 0; i < 2; ++i) {
+    if ((flags = fcntl(pipehack[i], F_GETFL)) == -1)
+      LOG(ERROR) << "get flags";
+    if (i == 1)
+      flags |= O_NONBLOCK;
+    if (fcntl(pipehack[i], F_SETFL, flags) == -1)
+      LOG(ERROR) << "set flags";
+    if ((flags = fcntl(pipehack[i], F_GETFD)) == -1)
+      LOG(ERROR) << "get fd flags";
+    flags |= FD_CLOEXEC;
+    if (fcntl(pipehack[i], F_SETFD, flags) == -1)
+      LOG(ERROR) << "set fd flags";
+  }
+
+  struct sigaction action;
+  memset(&action, 0, sizeof(action));
+  action.sa_handler = SIGCHLDHandler;
+  sigaction(SIGCHLD, &action, NULL);
+}
+#endif
+
+namespace brave {
+
+class TorLauncherDelegate : public base::LaunchOptions::PreExecDelegate {
+ public:
+    TorLauncherDelegate() {}
+    ~TorLauncherDelegate() override {}
+
+    void RunAsyncSafe() override {
+#if defined(OS_POSIX)
+      for (size_t i = 0; i < 2; ++i)
+        close(pipehack[i]);
+      setsid();
```

Comment:
Can you add a comment explaining why we do this?  Same as the answer you gave to an earlier review, basically.

Reply:
fixed in 4e72ec6